homeassistant/helpers/device_registry.py

```diff
@@ -349,21 +416,38 @@ def _data_to_save(self) -> Dict[str, List[Dict[str, Any]]]:
             }
             for entry in self.devices.values()
         ]
+        data["deleted_devices"] = [
+            {
+                "config_entries": list(entry.config_entries),
+                "connections": list(entry.connections),
+                "identifiers": list(entry.identifiers),
+                "id": entry.id,
+            }
+            for entry in self.deleted_devices.values()
+        ]
 
         return data
 
     @callback
     def async_clear_config_entry(self, config_entry_id: str) -> None:
         """Clear config entry from registry entries."""
         for device in list(self.devices.values()):
-            self._async_update_device(device.id, remove_config_entry_id=config_entry_id)
+            self._async_update_device(
+                self.devices, device.id, remove_config_entry_id=config_entry_id
+            )
+        for deleted_device in list(self.deleted_devices.values()):
+            self._async_update_device(
```

Comment:
We should not give `_async_update_device` this responsibility. We can just include a small for-loop in this method to pop entries related to this config entry.

Reply:
Agreed and rewritten accordingly.